conjure-core/src/main/java/com/palantir/conjure/parser/types/TypeParser.java

```diff
@@ -90,8 +90,15 @@ public LocalReferenceType parse(ParserState input) throws ParseException {
                 input.rewind();
                 return null;
             }
+            TypeName typeName;
+            try {
+                typeName = TypeName.of(typeReference);
+            } catch (IllegalArgumentException _e) {
+                input.rewind();
+                return null;
+            }
             input.release();
-            return LocalReferenceType.of(TypeName.of(typeReference));
```

Comment:
The negative test case was getting the wrong exception because the parsing was trapped here, which was how I ran into this bug in the first place, getting a cryptic and totally irrelevant exception.

Reply:
Makes sense!

For my own reference of why this is needed in `LocalReferenceTypeParser` and not `ForeignReferenceTypeParser`, this is because we go to the next parser if the previous one returned `null` and `ForeignReferenceTypeParser` is before `LocalReferenceTypeParser`:

https://github.com/palantir/conjure/blob/7ecb0ea62f0717e57b534943ce0dae4ad65c27de/conjure-core/src/main/java/com/palantir/conjure/parser/types/TypeParser.java#L58-L67